hydrachain/native_contracts.py

```diff
@@ -714,9 +728,33 @@ def __getitem__(self, i):
     def __setitem__(self, i, v):
         i = int(i)
         assert isinstance(i, (int, long))
+        if isinstance(self._value_type, Scalar):
+            if type(v).__name__ not in self._value_type._value_type:
+                raise TypeError("Value must be of a type " + self._value_type._value_type
+                                + ". Provided value of a type " + type(v).__name__ + " instead.")
+        else:
+            if type(v).__name__ not in self._value_type:
+                raise TypeError("Value must be of a type " + self._value_type
+                                + ". Provided value of a type " + type(v).__name__ + " instead.")
         self.set(bytes(i), v)
+        self.updatelen(i, v)
+
+    def updatelen(self, i, v):
         if i >= len(self):
             self.set(b'__len__', i + 1, value_type='uint32')
+        elif i+1 == len(self) and v is 0:
+            self.set(b'__len__', 0, value_type='uint32')
+            # there is no good way to recalculate this, so iterate for now, but think of a better solution
+            for k in reversed(range(i)):
```

Comment:
This change was reverted
https://github.com/RomanZacharia/hydrachain/blob/838029b53e1b29063296e1264c1002808b2b5fed/hydrachain/native_contracts.py#L744